File: packages/react-native-gtkx/tests/gtk/support/virtual-touchpad.py
#!/usr/bin/env python3
"""A virtual multitouch touchpad on /dev/uinput, driven over stdin.

The other half of ./virtual-touchpad.ts, which is the file to read first.
This half exists in Python because creating a uinput device needs ioctls
(UI_SET_EVBIT, UI_DEV_SETUP, ...) and Node has no ioctl; python3-evdev has
them, and it is the same technique libinput's own litest suite uses.

The device shape is an ordinary laptop clickpad: 100x60mm at 40 units/mm,
two MT slots, INPUT_PROP_POINTER. That is the whole of what libinput needs
to classify it as a touchpad (`cap:pg` — pointer and GESTURE) and start
running its pinch/swipe detection over it.

Protocol: one command per line on stdin, one `ok` per command on stdout once
the injected sequence has been written and flushed. `ready` is printed once,
after the device has been created and udev has had time to see it.

    pinch <scale> [steps]     two fingers, distance multiplied by <scale>
    rotate <degrees> [steps]  two fingers, rotated about their midpoint
    glide <dx_mm> <dy_mm>     one finger, which is how the pointer is moved
    quit
"""

import math
import sys
import time

from evdev import AbsInfo, UInput
from evdev import ecodes as e

# 100mm x 60mm at 40 units/mm. The resolution matters: libinput reports a
# touchpad without one as having no dimensions and falls back to a default,
# and its gesture thresholds are all in millimetres.
RESOLUTION = 40
X_MAX = 100 * RESOLUTION
Y_MAX = 60 * RESOLUTION

# BTN_TOOL_DOUBLETAP is what tells the kernel and libinput that two fingers
# are down; without it a two-slot report is not a two-finger gesture.
CAPABILITIES = {
    e.EV_KEY: [
        e.BTN_LEFT,
        # BTN_RIGHT is not decoration: without it libinput logs "kernel bug:
        # missing right button, assuming it is a clickpad" and reclassifies
        # the device. A touchpad with physical buttons is the simpler shape.
        e.BTN_RIGHT,
        e.BTN_TOUCH,
        e.BTN_TOOL_FINGER,
        e.BTN_TOOL_DOUBLETAP,
        e.BTN_TOOL_TRIPLETAP,
        e.BTN_TOOL_QUADTAP,
    ],
    e.EV_ABS: [
        (e.ABS_X, AbsInfo(0, 0, X_MAX, 0, 0, RESOLUTION)),
        (e.ABS_Y, AbsInfo(0, 0, Y_MAX, 0, 0, RESOLUTION)),
        (e.ABS_PRESSURE, AbsInfo(0, 0, 255, 0, 0, 0)),
        (e.ABS_MT_SLOT, AbsInfo(0, 0, 1, 0, 0, 0)),
        (e.ABS_MT_POSITION_X, AbsInfo(0, 0, X_MAX, 0, 0, RESOLUTION)),
        (e.ABS_MT_POSITION_Y, AbsInfo(0, 0, Y_MAX, 0, 0, RESOLUTION)),
        (e.ABS_MT_TRACKING_ID, AbsInfo(0, 0, 65535, 0, 0, 0)),
        (e.ABS_MT_PRESSURE, AbsInfo(0, 0, 255, 0, 0, 0)),
    ],
}

CENTRE = (X_MAX // 2, Y_MAX // 2)
# Frames arrive at roughly this interval on a real touchpad, and the interval
# is not decoration: libinput decides pinch-versus-scroll from the first few
# frames of motion and computes its velocities from the timestamps.
FRAME_MS = 12
PRESSURE = 60
# 15mm from the midpoint — 30mm apart, which is a comfortable two-finger
# starting span and leaves room to spread without running off the edge.
START_RADIUS = 15 * RESOLUTION


class Touchpad:
    def __init__(self):
        self.ui = UInput(
            CAPABILITIES,
            name="rn-gtkx virtual touchpad",
            vendor=0x0002,
            product=0x0007,
            version=0x01B1,
            phys="rn-gtkx/serio0/input0",
            input_props=[e.INPUT_PROP_POINTER],
        )
        self.tracking = 1

    def sync(self):
        self.ui.syn()
        time.sleep(FRAME_MS / 1000)

    def place(self, index, x, y):
        self.ui.write(e.EV_ABS, e.ABS_MT_SLOT, index)
        self.ui.write(e.EV_ABS, e.ABS_MT_POSITION_X, int(x))
        self.ui.write(e.EV_ABS, e.ABS_MT_POSITION_Y, int(y))
        self.ui.write(e.EV_ABS, e.ABS_MT_PRESSURE, PRESSURE)

    def down(self, points):
        """Puts every finger down in one frame."""
        for index, (x, y) in enumerate(points):
            self.ui.write(e.EV_ABS, e.ABS_MT_SLOT, index)
            self.ui.write(e.EV_ABS, e.ABS_MT_TRACKING_ID, self.tracking)
            self.tracking += 1
            self.place(index, x, y)
        self.ui.write(e.EV_KEY, e.BTN_TOUCH, 1)
        self.ui.write(e.EV_KEY, e.BTN_TOOL_FINGER, 1 if len(points) == 1 else 0)
        self.ui.write(e.EV_KEY, e.BTN_TOOL_DOUBLETAP, 1 if len(points) == 2 else 0)
        # Single-touch emulation follows the first finger. libinput reads the
        # MT slots, but a device that reports slots and no ABS_X/ABS_Y at all
        # is not a shape any driver produces.
        self.ui.write(e.EV_ABS, e.ABS_X, int(points[0][0]))
        self.ui.write(e.EV_ABS, e.ABS_Y, int(points[0][1]))
        self.ui.write(e.EV_ABS, e.ABS_PRESSURE, PRESSURE)
        self.sync()

    def move(self, points):
        for index, (x, y) in enumerate(points):
            self.place(index, x, y)
        self.ui.write(e.EV_ABS, e.ABS_X, int(points[0][0]))
        self.ui.write(e.EV_ABS, e.ABS_Y, int(points[0][1]))
        self.sync()

    def up(self, count):
        for index in range(count):
            self.ui.write(e.EV_ABS, e.ABS_MT_SLOT, index)
            self.ui.write(e.EV_ABS, e.ABS_MT_TRACKING_ID, -1)
        self.ui.write(e.EV_KEY, e.BTN_TOUCH, 0)
        self.ui.write(e.EV_KEY, e.BTN_TOOL_FINGER, 0)
        self.ui.write(e.EV_KEY, e.BTN_TOOL_DOUBLETAP, 0)
        self.ui.write(e.EV_ABS, e.ABS_PRESSURE, 0)
        self.sync()

    def two_finger(self, radii, angles):
        """Two fingers through a (radius, angle) path about the midpoint."""
        cx, cy = CENTRE

        def points(radius, angle):
            dx = radius * math.cos(angle)
            dy = radius * math.sin(angle)
            return [(cx - dx, cy - dy), (cx + dx, cy + dy)]

        self.down(points(radii[0], angles[0]))
        # No dwell before the motion: a real pinch starts moving at once, and
        # a pair of fingers that sits still for 150ms is what libinput's own
        # scroll timeout is waiting for.
        for radius, angle in zip(radii[1:], angles[1:]):
            self.move(points(radius, angle))
        self.up(2)

    def pinch(self, scale, steps):
        end = START_RADIUS * scale
        radii = [
            START_RADIUS + (end - START_RADIUS) * i / steps for i in range(steps + 1)
        ]
        self.two_finger(radii, [0.0] * (steps + 1))

    def rotate(self, degrees, steps):
        total = math.radians(degrees)
        angles = [total * i / steps for i in range(steps + 1)]
        self.two_finger([START_RADIUS] * (steps + 1), angles)

    def glide(self, dx_mm, dy_mm):
        cx, cy = CENTRE
        target_x = cx + dx_mm * RESOLUTION
        target_y = cy + dy_mm * RESOLUTION
        # One step per 3mm, floor 10. libinput discards a frame that moves a
        # finger further than its jump threshold ("Touch jump detected and
        # discarded"), so a long glide in ten steps silently loses travel.
        steps = max(10, math.ceil(math.hypot(dx_mm, dy_mm) / 3))
        self.down([(cx, cy)])
        for i in range(1, steps + 1):
            self.move(
                [
                    (
                        cx + (target_x - cx) * i / steps,
                        cy + (target_y - cy) * i / steps,
                    )
                ]
            )
        self.up(1)


def main():
    pad = Touchpad()
    # udev has to see the device and the compositor's libinput context has to
    # add it to the seat before anything injected through it means anything.
    time.sleep(1.5)
    print("ready", flush=True)
    for line in sys.stdin:
        parts = line.split()
        if not parts or parts[0] == "quit":
            break
        command = parts[0]
        try:
            if command == "pinch":
                pad.pinch(float(parts[1]), int(parts[2]) if len(parts) > 2 else 20)
            elif command == "rotate":
                pad.rotate(float(parts[1]), int(parts[2]) if len(parts) > 2 else 20)
            elif command == "glide":
                pad.glide(float(parts[1]), float(parts[2]))
            else:
                print(f"error unknown command {command}", flush=True)
                continue
        except Exception as error:  # noqa: BLE001 - reported, not swallowed
            print(f"error {error}", flush=True)
            continue
        print("ok", flush=True)
    pad.ui.close()


if __name__ == "__main__":
    main()

File: packages/react-native-gtkx/tests/gtk/support/virtual-touchpad.ts
// Real touchpad input, for the two gestures the pointer stream cannot carry.
//
// Why this exists, and why it is not `virtual-pointer.ts` with more requests:
// `zwlr_virtual_pointer_v1` has motion, button, axis and frame, and no
// gesture requests of any kind. There is no Wayland protocol for injecting a
// pinch — not in wlroots, not anywhere — because a pinch is not something a
// client synthesizes. It is something libinput CONCLUDES from two fingers
// moving apart on a device it has classified as a touchpad.
//
// So the injection point is one layer lower than the pointer's: a virtual
// multitouch touchpad on `/dev/uinput`, which is the technique libinput's own
// litest suite uses. The chain is then entirely real:
//
//   uinput -> kernel evdev -> libinput (GESTURE_PINCH_BEGIN/UPDATE/END)
//          -> compositor (zwp_pointer_gestures_v1)
//          -> GDK (GDK_TOUCHPAD_PINCH)
//          -> GtkGestureZoom / GtkGestureRotate
//
// MEASURED, and this is the constraint every caller has to know about: the
// chain works under a compositor with a libinput backend, and delivers
// nothing at all under one without. The headless sway `@gtkx/vitest` starts
// per worker is the second kind — it is launched with `WLR_BACKENDS=headless`
// and `WLR_LIBINPUT_NO_DEVICES=1`, so it enumerates no input devices and a
// uinput touchpad is invisible to it. Under the VM's real GNOME session the
// same injection reaches `GtkGestureZoom` with the right numbers.
// `spike/gesture-detector/run-session.sh` is what runs against the real one;
// see docs/research/gesture-detector.md, probe 6.
//
// The device itself is created from Python, in `./virtual-touchpad.py`:
// uinput needs ioctls and Node has none. This half is the process and the
// line protocol, in the shape `createVirtualPointer` already established, so
// a test drives a touchpad the same way it drives a pointer.
import {
  spawn,
  spawnSync,
  type ChildProcessWithoutNullStreams,
} from "node:child_process"
import { existsSync } from "node:fs"
import { dirname, join } from "node:path"
import { createInterface, type Interface } from "node:readline"

/** Where the helper sits in the repo, from the repo root. */
const HELPER_FROM_ROOT = join(
  "packages",
  "react-native-gtkx",
  "tests",
  "gtk",
  "support",
  "virtual-touchpad.py",
)

/**
 * The sibling `.py`, or the repo copy of it.
 *
 * A test running from source finds it next to this file, which is the whole of
 * it. A BUNDLED consumer does not: `spike/gesture-detector` builds this module
 * into `dist/bundle.mjs`, where `import.meta.dirname` is the bundle's directory
 * and there is no sibling to find — the helper is a Python file, so no bundler
 * will ever carry it along. Walking up for the repo path covers that without
 * an environment variable nobody would remember to set.
 */
const resolveHelper = (): string | null => {
  const sibling = join(import.meta.dirname, "virtual-touchpad.py")
  if (existsSync(sibling)) {
    return sibling
  }
  for (const start of [import.meta.dirname, process.cwd()]) {
    let directory = start
    for (;;) {
      const candidate = join(directory, HELPER_FROM_ROOT)
      if (existsSync(candidate)) {
        return candidate
      }
      const parent = dirname(directory)
      if (parent === directory) {
        break
      }
      directory = parent
    }
  }
  return null
}

const HELPER = resolveHelper()

/** How long a single injected sequence may take before it is given up on. */
const COMMAND_TIMEOUT_MS = 15_000

export class VirtualTouchpadUnavailable extends Error {}

export type VirtualTouchpad = {
  /**
   * Two fingers, their distance multiplied by `scale` over `steps` frames.
   * Above 1 spreads (zoom in), below 1 pinches.
   */
  pinchBy(scale: number, steps?: number): Promise<void>
  /** Two fingers rotated about their midpoint; positive is clockwise. */
  rotateBy(degrees: number, steps?: number): Promise<void>
  /**
   * One finger, in millimetres of touchpad travel — which is how the POINTER
   * is moved, since a touchpad has no absolute addressing. Pointer
   * acceleration sits between this and the pixels the pointer ends up moving,
   * so it aims rather than places: glide hard into a corner first (the
   * compositor clamps, so that lands exactly), then walk out from there.
   */
  glideBy(millimetresX: number, millimetresY: number): Promise<void>
  dispose(): void
}

/**
 * Everything that has to be true before a virtual touchpad can exist, checked
 * up front so a test can skip on an environment difference rather than fail
 * on one. Returns the reason it cannot, or null.
 */
export const virtualTouchpadBlocker = (): string | null => {
  if (process.platform !== "linux") {
    return "not Linux"
  }
  if (!existsSync("/dev/uinput")) {
    return "no /dev/uinput — the kernel module is not loaded"
  }
  if (HELPER === null) {
    return `the uinput helper is missing (${HELPER_FROM_ROOT})`
  }
  // Root, because /dev/uinput is root:input 0660 and the test user is not in
  // the `input` group on any machine this has run on.
  if (spawnSync("sudo", ["-n", "true"]).status !== 0) {
    return "no passwordless sudo, and /dev/uinput is root:input 0660"
  }
  if (spawnSync("python3", ["-c", "import evdev"]).status !== 0) {
    return "python3-evdev is not importable"
  }
  return null
}

type Pending = {
  resolve: () => void
  reject: (error: Error) => void
  timer: ReturnType<typeof setTimeout>
}

const settle = (pending: Pending[], line: string): void => {
  const next = pending.shift()
  if (!next) {
    return
  }
  clearTimeout(next.timer)
  if (line.startsWith("error")) {
    next.reject(new Error(`virtual touchpad: ${line}`))
    return
  }
  next.resolve()
}

/**
 * Creates the device and waits for udev to have seen it.
 *
 * Throws {@link VirtualTouchpadUnavailable} when the machine cannot host one,
 * so a test can skip rather than fail on an environment difference — the same
 * contract `createVirtualPointer` has for a compositor without
 * `zwlr_virtual_pointer_manager_v1`.
 */
export const createVirtualTouchpad = async (): Promise<VirtualTouchpad> => {
  const blocker = virtualTouchpadBlocker()
  if (blocker !== null) {
    throw new VirtualTouchpadUnavailable(blocker)
  }

  const child: ChildProcessWithoutNullStreams = spawn(
    "sudo",
    // `blocker` is null, so the helper was found.
    ["-n", "python3", HELPER as string],
    { stdio: ["pipe", "pipe", "pipe"] },
  )
  const pending: Pending[] = []
  let stderr = ""
  child.stderr.setEncoding("utf8")
  child.stderr.on("data", (chunk: string) => {
    stderr += chunk
  })

  const lines: Interface = createInterface({ input: child.stdout })
  const ready = new Promise<void>((resolve, reject) => {
    const onExit = (): void => {
      reject(
        new VirtualTouchpadUnavailable(
          `the uinput helper exited before it was ready: ${stderr.trim()}`,
        ),
      )
    }
    child.once("exit", onExit)
    lines.once("line", (line: string) => {
      child.off("exit", onExit)
      if (line.trim() === "ready") {
        resolve()
        return
      }
      reject(new VirtualTouchpadUnavailable(`the uinput helper said: ${line}`))
    })
  })
  await ready

  lines.on("line", (line: string) => {
    settle(pending, line.trim())
  })
  child.on("exit", () => {
    while (pending.length > 0) {
      settle(pending, "error the uinput helper exited")
    }
  })

  const send = (command: string): Promise<void> =>
    new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        reject(new Error(`virtual touchpad: \`${command}\` did not finish`))
      }, COMMAND_TIMEOUT_MS)
      pending.push({ resolve, reject, timer })
      child.stdin.write(`${command}\n`)
    })

  return {
    pinchBy: (scale, steps = 20) => send(`pinch ${scale} ${steps}`),
    rotateBy: (degrees, steps = 20) => send(`rotate ${degrees} ${steps}`),
    glideBy: (millimetresX, millimetresY) =>
      send(`glide ${millimetresX} ${millimetresY}`),
    dispose() {
      for (const entry of pending) {
        clearTimeout(entry.timer)
      }
      pending.length = 0
      child.stdin.end("quit\n")
      lines.close()
      // The helper closes the uinput device on its way out, which is what
      // makes the kernel remove it. If it will not go, the device would
      // outlive the test run.
      const killer = setTimeout(() => {
        child.kill("SIGKILL")
      }, 2000)
      child.once("exit", () => {
        clearTimeout(killer)
      })
    },
  }
}
